Sign NuGet Packages

rprouse opened the pull request:
Fixes #691 

This is the first pass at signing the NuGet packages. We could also sign the WiX installer, the Choco packages and the ZIP file if we want, but we'll need to make sure everything is installed on the build servers if we want to do the signing there. Right now, it just signs the NuGet packages, but I can easily extend it.

As I see it, we currently have a couple of options.

1. If we want to continue to build and release on the developer machines, I can extend this to sign everything. I can share the environment variables that need to be set with @ChrisMaddock so he can do a release.
2. If we want to build, sign and release from Azure DevOps, I need to do the following;
    - Extend the Cake script to sign everything
    - Figure out how to install WiX on the build machines
    - Setup the Azure DevOps build to build and sign the release branch?

Either way, I should probably extend the script to sign everything so I'll do that. I just wanted to get this out for review. 

@ChrisMaddock if you want to do a release soon, the first option is probably easiest then we can work on the second option as a second PR. What do you think?

At the moment, I am running the command `.\build.ps1 -Target SignNuGet` (may change the target to Sign) and the target outputs the following;

```
========================================
SignNuGet
========================================
Signing C:/Src/nunit/nunit-console/package/NUnit.Console.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\NUnit.Console.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\NUnit.Console.3.11.0.nupkg' in 3551 ms
Signing C:/Src/nunit/nunit-console/package/NUnit.ConsoleRunner.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\NUnit.ConsoleRunner.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\NUnit.ConsoleRunner.3.11.0.nupkg' in 3895 ms
Signing C:/Src/nunit/nunit-console/package/NUnit.Engine.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\NUnit.Engine.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\NUnit.Engine.3.11.0.nupkg' in 3962 ms
Signing C:/Src/nunit/nunit-console/package/NUnit.Engine.Api.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\NUnit.Engine.Api.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\NUnit.Engine.Api.3.11.0.nupkg' in 2317 ms
Signing C:/Src/nunit/nunit-console/package/nunit.engine.netstandard.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\nunit.engine.netstandard.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\nunit.engine.netstandard.3.11.0.nupkg' in 2096 ms
Signing C:/Src/nunit/nunit-console/package/NUnit.Runners.3.11.0.nupkg...
Submitting 'C:\Src\nunit\nunit-console\package\NUnit.Runners.3.11.0.nupkg' for signing.
Successfully signed 'C:\Src\nunit\nunit-console\package\NUnit.Runners.3.11.0.nupkg' in 1965 ms
```

rprouse commented:
This also updates Cake to the latest version.

rprouse commented:
I've pushed changes to sign the Chocolatey packages and the MSI installer.  Run now with `.\build.ps1 -Target SignPackages`

The MSI installer now shows the following in Properties;

![image](https://user-images.githubusercontent.com/493828/69769632-9a7c0c00-1153-11ea-807e-2e34705d08ac.png)

rprouse commented:
I've also confirmed that the signature is in the NuGet packages;

![image](https://user-images.githubusercontent.com/493828/69769794-563d3b80-1154-11ea-85a1-e0758477705f.png)

@ChrisMaddock do you want to go with this for the release? I can share the signing User and Password with you via LastPass.

ChrisMaddock commented:
Great! I agree - this’ll be good for the next release, and we can build on it and move it online in time. 😊

Just one note on option 2 - I believe the installer is already built on CI, so that’s one less task to tackle that one.

On my phone at the moment, but will have a proper look through the code soon.

rprouse commented:
Linux and MacOS builds are failing, probably because of upgrading Cake. I'll look into it.

jnm2 commented on build.cake:
```diff
@@ -1,5 +1,8 @@
 #load ci.cake
 
+// Install Tools
+#tool "nuget:https://api.nuget.org/v3/index.json?package=nuget.commandline&version=5.3.1"
```
Option:

```suggestion
#tool NuGet.CommandLine&version=5.3.1
```

jnm2 commented on build.cake:
```diff
@@ -545,6 +548,62 @@ Task("PackageZip")
+Task("InstallSigningTool")
+    .Does(() =>
+    {
+        var result = StartProcess("dotnet.exe", new ProcessSettings {  Arguments = "tool install SignClient -g" });
```
What do you think of installing to a local tool path instead of globally?
Also, how does this line behave when SignClient is already globally installed?

rprouse replied:
I think you're probably right, I originally did it that way. I did it this way because it is almost always run on a build machine. If it is already installed, the result code is non-zero, but I don't fail on that. I'll make the change to a local install.

jnm2 commented on build.cake:
```diff
@@ -545,6 +548,62 @@ Task("PackageZip")
+        //var filter = File("./signclient.filter");
+
+        // Get the files to sign.
+        var files = GetFiles(string.Concat(PACKAGE_DIR, "*.nupkg"));
```
Does SignClient also sign the .dll and .exe files within the nupkg?

rprouse replied:
No it doesn't sign the assemblies inside the package. Each file is submitted to be signed, so it can be a bit slow with too many things being signed. We could submit key assemblies for signing in the future though.

jnm2 commented on build.cake:
```diff
@@ -545,6 +548,62 @@ Task("PackageZip")
+                .AppendSwitchQuotedSecret("-s", secret)
+                .AppendSwitchQuotedSecret("-r", user)
+                .AppendSwitchQuoted("-n", "NUnit.org")
+                .AppendSwitchQuoted("-d", "NUnit is a unit-testing framework for all .Net languages.")
```
Branding:

```suggestion
                .AppendSwitchQuoted("-d", "NUnit is a unit-testing framework for all .NET languages.")
```

rprouse commented:
Sorry guys, I haven't been able to get back to this. I've torn out the shower in my house and we are going without until I build a new one. My wife not killing me takes priority 😄 

I HOPE the shower will be done this weekend and I can get back to this then. I'll keep you updated.

rprouse commented:
@jnm2 I've addressed your concerns. I've also fixed the Linux and MacOS build on AzureDevOps. Waiting for builds to complete to see if there is anything else that needs fixing up. Travis might too...

rprouse commented on azure-pipelines.yml:
```diff
@@ -59,7 +59,7 @@ jobs:
       sudo apt-get update
       sudo apt-get install dotnet-sdk-2.2
-      ./build.sh --target Test --configuration Release
+      ./build.sh --target=Test --configuration=Release
```
Apparently Cake cares that you use the format `--param=value` and no longer accepts `--param value`. We may need to change this in other projects if we update Cake.

rprouse commented:
Travis is still failing with less than useful errors after several attempts to fix it. It is now installing the same packages on Linux as Azure DevOps and calling the build script in the same way. @ChrisMaddock do you think Travis still provides value? I'd say the Linux build on Azure DevOps is enough. We've already disabled the Travis Mac builds because they are so flaky.

Other than the Travis failure, this is ready for review.

ChrisMaddock commented:
Happy with that - our original plan with this was to test out ADO as an alternative to travis, right? Would you mind just double checking ADO is definitely running everything travis is, and making the correct builds ‘required’ in github? I can’t quite remember where we got up to with it.

jnm2 commented on build.cake:
```diff
@@ -545,6 +548,61 @@ Task("PackageZip")
+            {
+                // We should not recover from this.
+                throw new InvalidOperationException("Signing failed!");
+            }
```
`DotNetCoreTool` is an option: https://cakebuild.net/api/Cake.Common.Tools.DotNetCore/DotNetCoreAliases/0C29BCC5

ChrisMaddock commented:
Hey @rprouse - just thought I'd try giving updating mono a try, and that seems to get Travis passing...so I've pushed a commit to do that. Hope you don't mind.

Happy enough to remove Travis, but this seemed an easier change to make for now, just to get this PR merged. We can remove Travis as a separate thing, if we still want. 🙂

ChrisMaddock commented:
And...merging. Thanks for sorting this one, Rob! 😁
